src/contexts/ActivityContext.js

```diff
@@ -0,0 +1,248 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import StoredList from '../StoredList'
+import { network } from '../environment'
+import { EthereumAddressType } from '../prop-types'
+
+const ActivityContext = React.createContext()
+
+const activityStatusTypes = {
+  CONFIRMED: 'CONFIRMED',
+  PENDING: 'PENDING',
+  FAILED: 'FAILED',
+  TIMED_OUT: 'TIMED_OUT',
+}
+
+const activityTypes = {
+  TRANSACTION: 'TRANSACTION',
+}
+
+const TEN_MINUTES = 1000 * 60 * 10
+
+let storedList
+
+// Provides easy access to the user activities list
+class ActivityProvider extends React.Component {
+  static propTypes = {
+    account: EthereumAddressType, // Current wallet
+    children: PropTypes.node,
+    daoDomain: PropTypes.string, // domain of current DAO
+    walletWeb3: PropTypes.object,
+  }
+  state = {
+    // activities of all accounts
+    activities: [],
+  }
+  static defaultProps = {
+    account: '',
+  }
+
+  componentDidUpdate(prevProps) {
+    const { daoDomain, account } = this.props
+
+    if (daoDomain !== prevProps.daoDomain || account !== prevProps.account) {
+      const storageKey = `activity:${network.type}:${daoDomain}:${account}`
+      storedList = new StoredList(storageKey)
+
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState(
+        { activities: storedList.getItems() },
+        this.refreshPendingActivities
+      )
+    }
+  }
+
+  componentWillMount() {
+    this._checkInterval = setInterval(this.checkForTimedOut, 1000 * 30)
+  }
+  componentWillUnmount() {
+    clearInterval(this._checkInterval)
+  }
+
+  // Refresh the status of pending activities
+  refreshPendingActivities() {
+    const { walletWeb3 } = this.props
+
+    this.state.activities
+      .filter(({ status }) => status === activityStatusTypes.PENDING)
+      .forEach(({ transactionHash }) => {
+        walletWeb3.eth
+          .getTransaction(transactionHash)
+          .then(({ blockNumber }) =>
+            blockNumber ? this.setActivityConfirmed(transactionHash) : null
+          )
+          .catch(console.error)
+      })
+  }
+
+  checkForTimedOut = () => {
+    const now = Date.now()
+    this.state.activities.forEach(activity => {
+      const timeDelta = now - activity.createdAt
+      if (
+        timeDelta > TEN_MINUTES &&
+        activity.status === activityStatusTypes.PENDING
+      ) {
+        // Set pending items to timed out after 10 minutes
+        this.setActivityTimedOut(activity.transactionHash)
+      }
+    })
+  }
+
+  add = activity => {
```

Comment:
Is this needed elsewhere? Could we always use `addTransactionActivity()`?

Reply:
Nope